# Skills Test 2

Welcome to your second skills test!

You may turn it in within the ipython notebook by saving the notebook with your name and emailing the notebook to the primary instructor and DSR.

### Part 1:

Recall that we write the linear equation $A\vec{x} = b$. When we describe a line in $\mathbb{R}^n$ (from the linear equation) in parametric form, we include the origin of the line in the equation, such that $\vec{x} = s\vec{u}+\vec{b}$, for $s \in \mathbb{R}$, where $\vec{u} \in \mathbb{R}^n$. 

You are given the linear system characterized by the matrix $A$ (given in RREF) and augmented by $b$:

$$A_{RREF}|b = \left [ \begin{array}{cccc|c}1 & 0 & 3 & 0 & 5\\ 0 & 1 & -2 & 0 & 3 \\ 0 & 0 & 0 & 1 & 1  \end{array} \right ]$$

1. You have solved many equations like this before. Here it is simply a linear system. What is the parametric equation of this linear system? What does the parametric equation describe?

2. If $A$ were a transformation $T$, what fields are the domain and range of the transformation? 

3. Given the above information, find all solutions to $A\vec{x} = 0$.

4. Does the matrix equation $A\vec{x} = c$ have a solution for every $c$ in $\mathbb{R}^3$? Provide a one sentence explanation for your answer.

5. Find $M^{9876543210}$ where $M = \begin{bmatrix}1&0\\2&-1\end{bmatrix}$

6. Consider the $2 \times 2$ matrix $B = \begin{bmatrix}a&b\\c&d\end{bmatrix}$ where $a,b,c,d$ are real numbers and $a \neq 0$. Find all values $d$ such that $rank(B) = 1$

7. Find a basis for the kernel of $A$ and a basis for the image of $A$, where $A = \begin{bmatrix}2&4&8\\4&5&1\\7&9&3\end{bmatrix}$

8. Find a basis for the space of all matrices $A = \begin{bmatrix}a&b\\c&d\end{bmatrix}$ where $a+b+c+d = 0$


### Part 2:


Given points $A = \begin{bmatrix}1 \\ 2 \\ -3\end{bmatrix}$, $B = \begin{bmatrix}1 \\ 5 \\ 0\end{bmatrix}$, $C = \begin{bmatrix}5 \\ 6 \\ -1\end{bmatrix}$ (problems 1, 2, 3, 4)

1. Show that $\vec{AB}$ is perpendicular to $\vec{BC}$ and find the area of the triangle $ABC$

2. Find the equation of the plane ABC.

3. Find the distance of the point $D = \begin{bmatrix}4 \\ -1 \\ 3\end{bmatrix}$ to $ABC$.

4. Show that the volume of tetrahedron $ABCD$ is 21.

5. Calculate the angle between the two planes $\begin{bmatrix}1 & -1 & 5\end{bmatrix}$ $\begin{bmatrix}x_{1} \\ x_{2} \\ X_{3}\end{bmatrix}=2$ and $\begin{bmatrix}3 & 2 & -1\end{bmatrix}$ $\begin{bmatrix}x_{1} \\ x_{2} \\ x_{3}\end{bmatrix}=5$

6. Where does the line $\vec{x} = \begin{bmatrix}1 \\ 1 \\ 0\end{bmatrix} + t\begin{bmatrix}2 \\ 3 \\ 4\end{bmatrix}$ intersect the plane $\begin{bmatrix}2 & 1 & -1\end{bmatrix}$ $\begin{bmatrix}x_{1} \\ x_{2} \\ x_{3}\end{bmatrix}=0$?

# Task 1

In [1]:
import numpy as np

### 1

1 * x + 0 * y + 3 * z + 0 * w = 5

0 * x + 1 * y - 2 * z + 0 * w = 3

0 * x + 0 * y + 0 * z + 1 * w = 1

w = 1

y - 2z = 3

x + 3z = 5

then,

z = (y - 3)/2 and x = 5 - 3z

If we parametrize y = t:

z = (t - 3)/2

x = (19 - 3t)/2

$\begin{bmatrix}x \\ y \\ z\\ w\end{bmatrix} = \begin{bmatrix}(19 - 3t)/2 \\ t \\(t-3)/2 \\1\end{bmatrix} =  \begin{bmatrix}19/2 \\ 0 \\-3/2 \\1\end{bmatrix} + t\begin{bmatrix}-3/2 \\ 1 \\ 1/2 \\ 0\end{bmatrix}$

The parametric equation describes a linear transformation T, where T is called multiplication by A.

### 2

Domain:

$\vec{b} = \begin{bmatrix}5 \\ 3 \\1\end{bmatrix}$

The range of transformation are the columns of the matrix A (given in RREF form):

$A_{RREF}= \left [ \begin{array}{cccc|c}1 & 0 & 3 & 0 \\ 0 & 1 & -2 & 0  \\ 0 & 0 & 0 & 1  \end{array} \right ]$

### 3

$A\vec{x} = 0$

The augmented matrix is:

$\left [ \begin{array}{cccc|c}1 & 0 & 3 & 0 & 0\\ 0 & 1 & -2 & 0 & 0 \\ 0 & 0 & 0 & 1 & 0  \end{array} \right ]$

x + 3z = 0

y - 2z = 0

w = 0

Then, y = 2z. If we parametrize y = t, z = t/2

x = -3z, x = -3t/2

$\begin{bmatrix}x \\ y \\ z\\ w\end{bmatrix} = \begin{bmatrix}-3t/2 \\ t \\t/2 \\0\end{bmatrix} =  \begin{bmatrix}0 \\ 0 \\0 \\0\end{bmatrix} + t\begin{bmatrix}-3/2 \\ 1 \\ 1/2 \\ 0\end{bmatrix}$

### 4

$A\vec{x} = c$

Yes because the solution can be parametrized as we saw on 1 and $c \in \mathbb{R}$.

### 5

In [2]:
M = np.asarray([1,0,2,-1])
M = M.reshape(2,2)
M

array([[ 1,  0],
       [ 2, -1]])

In [3]:
eigenvalues, eigenvectors = np.linalg.eig(M)

In [4]:
eigenvalues

array([-1.,  1.])

In [5]:
eigenvectors

array([[ 0.        ,  0.70710678],
       [ 1.        ,  0.70710678]])

So, $\lambda_1 = 1, \lambda_2 = -1$

Eigenvector 1 = $\begin{bmatrix}1\\1\end{bmatrix}$, which normalized is: $\begin{bmatrix}0.70710678\\0.70710678\end{bmatrix}$

Eigenvector 2 = $\begin{bmatrix}0\\1\end{bmatrix}$

We know that:

M = PDP$^{-1}$, and

M$^k$ = PD$^k$P$^{-1}$

$P = \begin{bmatrix}0.70710678&0\\0.70710678&1\end{bmatrix}$

$D = \begin{bmatrix}1&0\\0&-1\end{bmatrix}$

In [6]:
P = np.asarray([1/2**0.5, 0, 1/2**0.5, 1])
P = P.reshape(2,2)
P

array([[ 0.70710678,  0.        ],
       [ 0.70710678,  1.        ]])

In [7]:
P_inverse = np.linalg.inv(P)
P_inverse

array([[ 1.41421356,  0.        ],
       [-1.        ,  1.        ]])

In [8]:
D = np.asarray([1,0,0,-1])
D = D.reshape(2,2)
D

array([[ 1,  0],
       [ 0, -1]])

In [9]:
M_power = P.dot(np.linalg.matrix_power(D,9876543210).dot(P_inverse))
M_power

array([[ 1.,  0.],
       [ 0.,  1.]])

### 6

$B = \begin{bmatrix}a&b\\c&d\end{bmatrix}$ where $a,b,c,d$ are real numbers and $a \neq 0$. 

If $rank(B) = 1$, then only 1 column of B in independent.

$x_1 = \begin{bmatrix}a\\c\end{bmatrix}$

$x_2 = \begin{bmatrix}b\\d\end{bmatrix}$

If $x_1$ is the independent column vector, then:

$x_2 = k * x_1$, where k is a scalar $k \in \mathbb{R}$.

$b = k * a$

$d = k * c$

Then, $k = b/a$ and $d = b*c/a$

### 7

In [10]:
K = np.asarray([2,4,8,4,5,1,7,9,3])
K = K.reshape(3,3)
K

array([[2, 4, 8],
       [4, 5, 1],
       [7, 9, 3]])

To find the basis of the kernel we need to solve the following system:

Kx = 0 (an homogeneous system), so the augmented matrix is:

$\begin{bmatrix}2&4&8&0\\4&5&1&0\\7&9&3&0\end{bmatrix}$

### 8

$A = \begin{bmatrix}a&b\\c&d\end{bmatrix} = a \begin{bmatrix}1&0\\0&0\end{bmatrix} + b \begin{bmatrix}0&1\\0&0\end{bmatrix} + c \begin{bmatrix}0&0\\1&0\end{bmatrix} + d \begin{bmatrix}0&0\\0&1\end{bmatrix}$

So the basis is:

$\begin{bmatrix}1&0\\0&0\end{bmatrix}, \begin{bmatrix}0&1\\0&0\end{bmatrix}, \begin{bmatrix}0&0\\1&0\end{bmatrix}, \begin{bmatrix}0&0\\0&1\end{bmatrix}$

# Task 2

### 1

In [11]:
a = np.asarray([1,2,-3])
b = np.asarray([1,5,0])
c = np.asarray([5,6,-1])
a = a.reshape(3,1)
b = b.reshape(3,1)
c = c.reshape(3,1)

In [12]:
# Creating a vector from points a, b.
a_b = a - b

In [13]:
a_b

array([[ 0],
       [-3],
       [-3]])

In [14]:
# Creating another vector from points b, c.
b_c = b - c

In [15]:
b_c

array([[-4],
       [-1],
       [ 1]])

In [16]:
# If vectors are perpendicular, then, their inner (dot) product is 0. The cosinus of 90 is 0.
a_b_c = a_b.T.dot(b_c)

In [17]:
a_b_c

array([[0]])

In [18]:
# Area: both vector are perpendicular. The area is the product of their norm's divided by 2 (base * height /2)
area = np.linalg.norm(a_b) * np.linalg.norm(b_c) / 2

In [19]:
area

8.9999999999999982

### 2

In [20]:
# Let's find the normal to the plane:
normal = np.cross(a_b.T,b_c.T)

In [21]:
normal = normal.T

In [22]:
normal

array([[ -6],
       [ 12],
       [-12]])

In [23]:
normal/6

array([[-1.],
       [ 2.],
       [-2.]])

The equation of a plane is:

ax + by + cz = d, a = -1, b = 2, c = -2

Using one point (1, 2, -3)

-1(x-1) + 2(y-2) - 2(z+3) = 0, -x + 1 + 2y - 4 - 2z - 6 = 0, -x + 2y - 2z = 6 + 4 - 1

Equation is:

-x + 2y - 2z = 9

Checking if point a is on the above plane:

In [24]:
-1 * 1 + 2 * 2 - 2 * (-3)

9

9 equal to 9, so equation is OK.

### 3

In [25]:
d = np.asarray([4,-1,3])
d = d.reshape(3,1)

In [26]:
# Make sure point is not on the plane:
-4 + 2 * (-1) - 2 * 3

-12

-12 does not equal to 9, point is not on the plane.

We know the normal to the plane is:

$\begin{bmatrix}-1 & 2 & -2\end{bmatrix}^T$

In [27]:
normal = np.asarray([-1,2,-2])
normal = normal.reshape(3,1)

Let's get a point on the plane, like (-1,0,-4).

We check if belongs to the plane:

$-1 * (-1) + 2 * 0 - 2 * (-4) = 1 + 8 = 9$, it belongs to the plane.

Creating a new vector:

In [28]:
e = np.asarray([-1,0,-4])
e = e.reshape(3,1)

In [29]:
d_e = d - e

In [30]:
d_e

array([[ 5],
       [-1],
       [ 7]])

Now, let's take the projection of the vector d_e to the normal vector;

In [31]:
distance = d_e.T.dot(normal)/(np.linalg.norm(normal))

In [32]:
distance

array([[-7.]])

If we take the absolute value, the distance is:

In [33]:
abs(distance)

array([[ 7.]])

### 4

The volume is the area of the base triangle times the height (which is the distance we calculated above), then divided by 3.
http://mathcentral.uregina.ca/qq/database/qq.09.03/peter2.html

In [52]:
volume = area * abs(distance)/3
volume

array([[ 21.]])

### 5

The equation of 1st plane is:

x - y + 5z = 2

The equation of the 2nd plane is:

3x + 2y - z = 5

In [43]:
# The normal to the 1st plane:
normal_plane_1 = np.asarray([1,-1,5])
normal_plane_1 = normal_plane_1.reshape(3,1)
normal_plane_1

array([[ 1],
       [-1],
       [ 5]])

In [42]:
abs(volume/4)

31.5

In [44]:
# The normal to the 2nd plane:
normal_plane_2 = np.asarray([3,2,-1])
normal_plane_2 = normal_plane_2.reshape(3,1)
normal_plane_2

array([[ 3],
       [ 2],
       [-1]])

We know that:

$||A|| * ||B|| * cos\theta = A * B$

In [45]:
normal_plane_1_norm = np.linalg.norm(normal_plane_1)
normal_plane_2_norm = np.linalg.norm(normal_plane_2)

In [46]:
cos_theta = normal_plane_1.T.dot(normal_plane_2)/(normal_plane_1_norm * normal_plane_2_norm)
cos_theta

array([[-0.2057378]])

Using a calculator to find $\theta = cos^{-1} (-0.2057378) = 101.87^o$

I was carried away and I also calculated the equation of the line(s) that intersects both planes:

I we put both equations in an augmented matrix, we get:

$\begin{bmatrix}1 & -1 & 5 & 2\\3 & 2 & -1 & 5\end{bmatrix}$

Performing row reduction, last row = last row - first row * 3:

$\begin{bmatrix}1 & -1 & 5 & 2\\0 & 5 & -16 & -1\end{bmatrix}$

5y - 16z = -1, y = (16z - 1)/5

If we parametrize z as t:

z = t

y = (16t - 1)/5

Then, x = 2 + y - 5z:

x = 2 - 5t + ((16t - 1)/5)

x = (9 - 9t)/5

$\begin{bmatrix}x\\y\\z\end{bmatrix} = \begin{bmatrix}(9 - 9t)/5\\(16t - 1)/5\\t\end{bmatrix} = \begin{bmatrix}9/5\\-1/5\\0\end{bmatrix} + t \begin{bmatrix}-9/5\\16/5\\1\end{bmatrix} = \frac{1}{5} * (\begin{bmatrix}9\\-1\\0\end{bmatrix} + t \begin{bmatrix}-9\\16\\1/5\end{bmatrix})$

### 6

From the line equation, we get:
    
x - 2t = 1

y - 3t = 1

z - 4t = 0

From the plane equation, we get:

2x - y - z = 0

Putting all equations in an matrix (the unknown coefficients are x, y, z and t):

2 * x - 1 * y - 1 * z - 0 * t = 0

1 * x + 0 * y + 0 * z - 2 * t = 1

0 * x + 1 * y + 0 * z - 3 * t = 1

0 * x + 0 * y + 1 * z - 4 * t = 0

In [47]:
Q = np.asarray([2,-1,-1,0,0,1,0,0,-2,1,0,1,0,-3,1,0,0,1,-4,0])
Q = Q.reshape(4,5)
# Q is the augmented matrix.
Q

array([[ 2, -1, -1,  0,  0],
       [ 1,  0,  0, -2,  1],
       [ 0,  1,  0, -3,  1],
       [ 0,  0,  1, -4,  0]])

In [48]:
Coeff = np.asarray([2,-1,-1,0,1,0,0,-2,0,1,0,-3,0,0,1,-4])
Coeff = Coeff.reshape(4,4)
Coeff

array([[ 2, -1, -1,  0],
       [ 1,  0,  0, -2],
       [ 0,  1,  0, -3],
       [ 0,  0,  1, -4]])

In [49]:
b = np.asarray([0,1,1,0])
b = b.reshape(4,1)
b

array([[0],
       [1],
       [1],
       [0]])

In [50]:
intersection = np.linalg.solve(Coeff,b)

In [51]:
intersection

array([[ 1.66666667],
       [ 2.        ],
       [ 1.33333333],
       [ 0.33333333]])

So, x = 1.67, y = 2, z = 1.33, t = 0.33